lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -0,0 +1,285 @@
+                    key,
+                    self._data[key],
+                    self._desired_config[key],
+                )
+                confict_keys.append(key)
+
+        return confict_keys
+
+    def _create_charm_file(self) -> None:
+        """Write the charm file."""
+        charm_params = [f"{key}={value}\n" for key, value in self._desired_config.items()]
+        with open(self.charm_filepath, "w") as f:
+            f.writelines(charm_params)
+
+    def _merge(self, add_own_charm=True) -> None:
+        """Create the merged sysctl file.
+
+        Args:
+            add_own_charm : bool, if false it will skip the charm file from the merge.
+        """
+        # get all files that start by 90-juju-
+        data = [SYSCTL_HEADER]
+        paths = set(SYSCTL_DIRECTORY.glob(f"{CHARM_FILENAME_PREFIX}*"))
+        if not add_own_charm:
+            paths.discard(self.charm_filepath.as_posix())
+
+        for path in paths:
+            with open(path, "r") as f:
+                data += f.readlines()
+        with open(SYSCTL_FILENAME, "w") as f:
+            f.writelines(data)
+
+        # Reload data with newly created file.
+        self._data = self._load_data()
+
+    def _apply(self) -> None:
+        """Apply values to machine."""
+        cmd = [f"{key}={value}" for key, value in self._desired_config.items()]
+        result = self._sysctl(cmd)
+        expr = re.compile(r"^sysctl: permission denied on key \"([a-z_\.]+)\", ignoring$")
+        failed_values = [expr.match(line) for line in result if expr.match(line)]
+        logger.debug("Failed values: %s", failed_values)
+
+        if failed_values:
+            msg = f"Unable to set params: {[f.group(1) for f in failed_values]}"
+            logger.error(msg)
+            raise SysctlPermissionError(msg)
+
+    def _create_snapshot(self) -> Dict[str, int]:
+        """Create a snaphot of config options that are going to be set."""
+        return {key: int(self._sysctl([key, "-n"])[0]) for key in self._desired_config.keys()}
+
+    def _restore_snapshot(self, snapshot: Dict[str, int]) -> None:
+        """Restore a snapshot to the machine."""
+        values = [f"{key}={value}" for key, value in snapshot.items()]
+        self._sysctl(values)
+
+    def _sysctl(self, cmd: List[str]) -> List[str]:
+        """Execute a sysctl command."""
+        cmd = ["sysctl"] + cmd
+        logger.debug("Executing sysctl command: %s", cmd)
+        try:
+            return check_output(cmd, stderr=STDOUT, universal_newlines=True).splitlines()
+        except CalledProcessError as e:
+            msg = f"Error executing '{cmd}': {e.stdout}"
+            logger.error(msg)
+            raise SysctlError(msg)
+
+    def _parse_config(self, config: Dict[str, dict]) -> None:
+        """Parse a config passed to the lib."""
+        result = {}
+        for key, value in config.items():
+            result[key] = int(value["value"])
```

Comment:
Yes, but than I saw `return {param.strip(): int(value.strip())}` in `_parse_line`, where we are doing conversion to integer also. If we will not do that, the validation could failed since `1 != "1"`. I'm afraid that we need to do some magic here to load values as int or sting.
```python
try:
    return int(value)
except ValueError:
    return value
```
But what about float, are there any possible configuration to be define as float?